### PR TITLE
Fastcov tool for code coverage

### PR DESCRIPTION
**Pull request by genadz**
## Tasklist

 - [ ] Add tests
 - [ ] Add #fixes with the issue number that this PR addresses
 - [ ] Generally use squash merge to rebase and clean comments before merging
 - [ ] Update the [changelog](CHANGELOG.md)

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -161,8 +161,6 @@ jobs:
       - run: scripts/clang-tidy-only-diff.sh 4
       - run: make -C build install
       - run: make -C build package
-      - run: make -C build coverage
```
Comment (PureW) [inline .circleci/config.yml:164]: There's been discussions about removing coverage checks but we haven't reached a conclusion yet.

Comment (PureW) [reply .circleci/config.yml:164]: The hope was rather that it would be possible to speed the process up, rather than removing it entirely.

Comment (mandeepsandhu) [reply .circleci/config.yml:164]: Since non one yet looks at coverage reports, can we make this workflow optional? (not sure if it already is)

**Comment by kevinkreiser** [comment]
currently the coverage check builds fail if coverage doesnt meet a certain threshold. we have ignored them in the past when they are very close (as they often are). additionally at least one of the coverage checks is misconfigured and should be removed. i think it reports on commit to commit coverage changes which arent useful. we only care about overall pr coverage change.

we do not want to disable coverage checks entirely. what we want is a way to continue to measure them but for it to not take so long. coverage checks keep us honest and they show us where we have written code that is untested.

in my opinion there are two things we need to look into: 

1. is the coverage check itself slow, if so make the check faster (fastcov?) by either checking/reporting on less code or by using a different tool
2. is the coverage upload to codecov.io slow? if so try cut out some of the report or use a different codecoverage report website

**Comment by genadz** [comment]
> currently the coverage check builds fail if coverage doesnt meet a certain threshold. we have ignored them in the past when they are very close (as they often are). additionally at least one of the coverage checks is misconfigured and should be removed. i think it reports on commit to commit coverage changes which arent useful. we only care about overall pr coverage change.
> 
> we do not want to disable coverage checks entirely. what we want is a way to continue to measure them but for it to not take so long. coverage checks keep us honest and they show us where we have written code that is untested.
> 
> in my opinion there are two things we need to look into:
> 
> 1. is the coverage check itself slow, if so make the check faster (fastcov?) by either checking/reporting on less code or by using a different tool
> 2. is the coverage upload to codecov.io slow? if so try cut out some of the report or use a different codecoverage report website

1. coverage check takes ~2-3 min and ~2-3 min takes to upload reports. `fastcov` requires gcc >= 9.0; I'd propose to try `fastcov` after we start using ubuntu 20.04 on CI

2. I've noticed that we include `test` folder in coverage report - imho it's needless information because we want to cover only business logic, so I'm going to remove files from `test` folder from report.

File: .circleci/config.yml [hunk 2]
```diff
@@ -94,6 +94,11 @@ jobs:
             - ccache-debug-linux-x86_64-{{ .Branch }}
             - ccache-debug-linux-x86_64
       - run: mkdir build
+      - run: |
+          # Directly download python script for coverage
+          cd build \
+          && curl https://raw.githubusercontent.com/RPGillespie6/fastcov/master/fastcov.py -o fastcov.py \
```
[the remark this answers is not in the document]
Comment (genadz) [reply .circleci/config.yml:100]: `fastcov` repository takes 444Kb. so, I think it's a good idea to have it as a thirdparty dep

File: CMakeLists.txt [hunk 3]
```diff
@@ -103,17 +103,16 @@ include(cmake/SanitizerOptions.cmake)
...
+  if(NOT FASTCOV_PATH OR NOT GENHTML_PATH)
+    message(FATAL_ERROR "no fastcov or genhtml installed")
   endif()
 
   add_custom_command(OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/coverage.info
-    COMMAND ${LCOV_PATH} --directory . --base-directory ${VALHALLA_SOURCE_DIR} --no-external --capture --output-file coverage.info --no-checksum
-    COMMAND ${LCOV_PATH} --remove coverage.info '*/third_party/*' --output-file coverage.info
-    COMMAND ${LCOV_PATH} --remove coverage.info '${CMAKE_CURRENT_BINARY_DIR}/*' --output-file coverage.info
-    COMMAND ${LCOV_PATH} --list coverage.info
+    COMMAND ${FASTCOV_PATH} -d . --exclude /usr/ third_party/ ${CMAKE_CURRENT_BINARY_DIR}/ --lcov -o coverage.info
```
[the remark this answers is not in the document]
Comment (genadz) [reply CMakeLists.txt:115]: by default `fastcov` runs in 8 threads

File: CMakeLists.txt [hunk 4]
```diff
@@ -103,17 +103,16 @@ include(cmake/SanitizerOptions.cmake)
 
 ## Coverage report targets
 if(ENABLE_COVERAGE)
-  find_program(LCOV_PATH NAMES lcov lcov.bat lcov.exe lcov.perl)
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -g -O0 --coverage")
```
[the remark this answers is not in the document]
Comment (genadz) [reply CMakeLists.txt:106]: hm, good catch. I missed the second place where `ENABLE_COVERAGE` is used: https://github.com/valhalla/valhalla/blob/master/src/CMakeLists.txt#L98-L100 . So, in such case it's not necessary to add coverage flags here. (`--coverage` includes `-ftest-coverage -fprofile-arcs` flags for compiler and `-lgcov` for linker)

[the remark this answers is not in the document]
Comment (genadz) [reply CMakeLists.txt:106]: about `-g -O0`. in release mode I didn't get errors, but result was a little bit different.

**Comment by genadz** [comment]
@kevinkreiser @PureW could you help me understand: how it's possible to get red flag for coverage job without code changing ? where can I find code coverage acceptance criterion ?

**Pull request merged**